packages/medusa/src/workflows/functions/attach-sales-channel-to-products.ts

```diff
@@ -0,0 +1,49 @@
+import { MedusaContainer, ProductTypes } from "@medusajs/types"
+import { EntityManager } from "typeorm"
+import { SalesChannelService } from "../../services"
+
+type ProductHandle = string
+type SalesChannelId = string
+
+export async function attachSalesChannelToProducts({
+  container,
+  manager,
+  data,
+}: {
+  container: MedusaContainer
+  manager: EntityManager
+  data: {
+    productsHandleSalesChannelsMap: Map<ProductHandle, SalesChannelId[]>
+    products: ProductTypes.ProductDTO[]
+  }
+}): Promise<void> {
+  const salesChannelService: SalesChannelService = container.resolve(
+    "salesChannelService"
+  )
+  const salesChannelServiceTx = salesChannelService.withTransaction(manager)
+
+  const salesChannelIdProductIdsMap = new Map<ProductHandle, SalesChannelId[]>()
+  data.products.forEach((product) => {
```

Comment:
General comment for all the compensation steps:
We have to use the optional chaining to get the data `data.products?.forEach` because if the workflow has failed at the correpondent invoke step, the information might not be available.

Reply:
I ve done the changes where it was making sense, let me know wyt